src/Compilers/CSharp/Portable/BoundTree/LengthBasedStringSwitchData.cs

```diff
@@ -0,0 +1,313 @@
+            {
+                // We pick the position that maximizes number of buckets with a single entry.
+                // We break ties by preferring lower max bucket size.
+                Debug.Assert(stringLength > 0);
+                Debug.Assert(caseLabels.Length > 0);
+                int bestIndex = -1;
+                int bestIndexSingleEntryCount = -1;
+                int bestIndexLargestBucket = int.MaxValue;
+                for (int currentPosition = 0; currentPosition < stringLength; currentPosition++)
+                {
+                    (int singleEntryCount, int largestBucket) = positionScore(currentPosition, caseLabels);
+
+                    if (singleEntryCount > bestIndexSingleEntryCount ||
+                        (singleEntryCount == bestIndexSingleEntryCount && largestBucket < bestIndexLargestBucket))
+                    {
+                        bestIndexSingleEntryCount = singleEntryCount;
+                        bestIndexLargestBucket = largestBucket;
+                        bestIndex = currentPosition;
+                    }
+                }
+
+                return bestIndex;
+            }
+
+            // Given a position and a set of string cases of matching lengths, inspect the buckets created by inspecting
+            // those strings at that position. Return the count how many buckets have a single entry and the size of the largest bucket.
+            static (int singleEntryCount, int largestBucket) positionScore(int position, ImmutableArray<(ConstantValue value, LabelSymbol label)> caseLabels)
+            {
+                var countPerChar = PooledDictionary<char, int>.GetInstance();
+                foreach (var caseLabel in caseLabels)
+                {
+                    Debug.Assert(caseLabel.value.StringValue is not null);
+                    var currentChar = caseLabel.value.StringValue[position];
+                    if (countPerChar.TryGetValue(currentChar, out var currentCount))
+                    {
+                        countPerChar[currentChar] = currentCount + 1;
+                    }
+                    else
+                    {
+                        countPerChar[currentChar] = 1;
+                    }
+                }
+
+                var singleEntryCount = countPerChar.Values.Count(c => c == 1);
+                var largestBucket = countPerChar.Values.Max();
+                countPerChar.Free();
+                return (singleEntryCount, largestBucket);
+            }
+        }
+
+        private static LabelSymbol CreateAndRegisterStringJumpTable(ImmutableArray<(ConstantValue value, LabelSymbol label)> cases, ArrayBuilder<StringJumpTable> stringJumpTables)
+        {
+            Debug.Assert(cases.Length > 0);
+
+            if (cases.Length == 1 && cases[0].value.StringValue!.Length == 1)
+            {
+                // If we have a single case that consists of a 1-length string then we can skip the string check
```

Comment:
>// If we have a single case that consists of a 1-length string then we can skip the string check [](http://example.com/codeflow?start=16&length=96)

Is this correct? What if the single character input string does not match?

Reply:
We only get here after a length and char matched. If the only case is a single character, that must be the case we're looking for. 
Doing a full string comparison would check all the same information again.